File: 2022/day19/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <regex>
#include <unordered_map>
#include <algorithm>

struct Blueprint {
    int id;
    int oreOre;
    int clayOre;
    int obsOre;
    int obsClay;
    int geoOre;
    int geoObs;
};

std::vector<Blueprint> parseInput(const std::string& filename) {
    std::vector<Blueprint> blueprints;
    std::ifstream file(filename);
    std::string line;

    std::regex pattern(R"(Blueprint (\d+): Each ore robot costs (\d+) ore\. Each clay robot costs (\d+) ore\. Each obsidian robot costs (\d+) ore and (\d+) clay\. Each geode robot costs (\d+) ore and (\d+) obsidian\.)");

    while (std::getline(file, line)) {
        std::smatch match;
        if (std::regex_search(line, match, pattern)) {
            Blueprint bp;
            bp.id = std::stoi(match[1]);
            bp.oreOre = std::stoi(match[2]);
            bp.clayOre = std::stoi(match[3]);
            bp.obsOre = std::stoi(match[4]);
            bp.obsClay = std::stoi(match[5]);
            bp.geoOre = std::stoi(match[6]);
            bp.geoObs = std::stoi(match[7]);
            blueprints.push_back(bp);
        }
    }
    return blueprints;
}

class GeodeSolver {
private:
    const Blueprint& bp;
    int timeLimit;
    int maxOre;
    int maxClay;
    int maxObs;
    int best;
    std::unordered_map<uint64_t, int> seen;

    uint64_t makeKey(int time, int ore, int clay, int obs, int oreR, int clayR, int obsR, int geoR) {
        // Pack state into 64-bit key
        return ((uint64_t)time << 56) |
               ((uint64_t)(ore & 0xFF) << 48) |
               ((uint64_t)(clay & 0xFF) << 40) |
               ((uint64_t)(obs & 0xFF) << 32) |
               ((uint64_t)(oreR & 0xFF) << 24) |
               ((uint64_t)(clayR & 0xFF) << 16) |
               ((uint64_t)(obsR & 0xFF) << 8) |
               ((uint64_t)(geoR & 0xFF));
    }

    void dfs(int time, int ore, int clay, int obs, int geodes,
             int oreR, int clayR, int obsR, int geoR) {
        // Pruning: upper bound on possible geodes
        int remaining = timeLimit - time;
        int upperBound = geodes + geoR * remaining + (remaining * (remaining - 1)) / 2;
        if (upperBound <= best) return;

        if (time == timeLimit) {
            best = std::max(best, geodes);
            return;
        }

        // Cap resources
        int cappedOre = std::min(ore, remaining * maxOre);
        int cappedClay = std::min(clay, remaining * maxClay);
        int cappedObs = std::min(obs, remaining * maxObs);

        // State deduplication
        uint64_t key = makeKey(time, cappedOre, cappedClay, cappedObs, oreR, clayR, obsR, geoR);
        auto it = seen.find(key);
        if (it != seen.end() && it->second >= geodes) return;
        seen[key] = geodes;

        // Collect resources
        int newOre = cappedOre + oreR;
        int newClay = cappedClay + clayR;
        int newObs = cappedObs + obsR;
        int newGeodes = geodes + geoR;

        // Try building geode robot (always do if possible)
        if (cappedOre >= bp.geoOre && cappedObs >= bp.geoObs) {
            dfs(time + 1, newOre - bp.geoOre, newClay, newObs - bp.geoObs, newGeodes,
                oreR, clayR, obsR, geoR + 1);
            return; // If we can build geode, always do
        }

        // Try building obsidian robot
        if (cappedOre >= bp.obsOre && cappedClay >= bp.obsClay && obsR < maxObs) {
            dfs(time + 1, newOre - bp.obsOre, newClay - bp.obsClay, newObs, newGeodes,
                oreR, clayR, obsR + 1, geoR);
        }

        // Try building clay robot
        if (cappedOre >= bp.clayOre && clayR < maxClay) {
            dfs(time + 1, newOre - bp.clayOre, newClay, newObs, newGeodes,
                oreR, clayR + 1, obsR, geoR);
        }

        // Try building ore robot
        if (cappedOre >= bp.oreOre && oreR < maxOre) {
            dfs(time + 1, newOre - bp.oreOre, newClay, newObs, newGeodes,
                oreR + 1, clayR, obsR, geoR);
        }

        // Do nothing (wait)
        dfs(time + 1, newOre, newClay, newObs, newGeodes,
            oreR, clayR, obsR, geoR);
    }

public:
    GeodeSolver(const Blueprint& blueprint, int limit)
        : bp(blueprint), timeLimit(limit), best(0) {
        maxOre = std::max({bp.oreOre, bp.clayOre, bp.obsOre, bp.geoOre});
        maxClay = bp.obsClay;
        maxObs = bp.geoObs;
    }

    int solve() {
        seen.clear();
        best = 0;
        dfs(0, 0, 0, 0, 0, 1, 0, 0, 0);
        return best;
    }
};

int part1(const std::vector<Blueprint>& blueprints) {
    int total = 0;
    for (const auto& bp : blueprints) {
        GeodeSolver solver(bp, 24);
        int geodes = solver.solve();
        total += bp.id * geodes;
    }
    return total;
}

int part2(const std::vector<Blueprint>& blueprints) {
    int result = 1;
    int count = std::min((int)blueprints.size(), 3);
    for (int i = 0; i < count; i++) {
        GeodeSolver solver(blueprints[i], 32);
        int geodes = solver.solve();
        result *= geodes;
    }
    return result;
}

int main() {
    auto blueprints = parseInput("../input.txt");

    std::cout << "Part 1: " << part1(blueprints) << std::endl;
    std::cout << "Part 2: " << part2(blueprints) << std::endl;

    return 0;
}

File: 2022/day19/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
)

type Blueprint struct {
	id      int
	oreOre  int
	clayOre int
	obsOre  int
	obsClay int
	geoOre  int
	geoObs  int
}

// State packed into uint64 for efficient map keys
// Bits: time(6) | ore(8) | clay(8) | obs(8) | oreR(6) | clayR(6) | obsR(6) | geoR(6)
type State struct {
	time, ore, clay, obs int
	oreR, clayR, obsR, geoR int
}

func (s State) key() uint64 {
	return uint64(s.time)<<58 |
		uint64(s.ore&0xFF)<<50 |
		uint64(s.clay&0xFF)<<42 |
		uint64(s.obs&0xFF)<<34 |
		uint64(s.oreR&0x3F)<<28 |
		uint64(s.clayR&0x3F)<<22 |
		uint64(s.obsR&0x3F)<<16 |
		uint64(s.geoR&0x3F)<<10
}

func parseInput(filename string) []Blueprint {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	pattern := regexp.MustCompile(`Blueprint (\d+): Each ore robot costs (\d+) ore\. Each clay robot costs (\d+) ore\. Each obsidian robot costs (\d+) ore and (\d+) clay\. Each geode robot costs (\d+) ore and (\d+) obsidian\.`)

	var blueprints []Blueprint
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		matches := pattern.FindStringSubmatch(line)
		if matches != nil {
			nums := make([]int, 7)
			for i := 0; i < 7; i++ {
				nums[i], _ = strconv.Atoi(matches[i+1])
			}
			blueprints = append(blueprints, Blueprint{
				id:      nums[0],
				oreOre:  nums[1],
				clayOre: nums[2],
				obsOre:  nums[3],
				obsClay: nums[4],
				geoOre:  nums[5],
				geoObs:  nums[6],
			})
		}
	}
	return blueprints
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxGeodes(bp Blueprint, timeLimit int) int {
	// Max robots needed per type
	maxOre := max(max(bp.oreOre, bp.clayOre), max(bp.obsOre, bp.geoOre))
	maxClay := bp.obsClay
	maxObs := bp.geoObs

	best := 0
	seen := make(map[uint64]int)

	var dfs func(time, ore, clay, obs, geodes, oreR, clayR, obsR, geoR int)
	dfs = func(time, ore, clay, obs, geodes, oreR, clayR, obsR, geoR int) {
		remaining := timeLimit - time

		// Pruning: upper bound on possible geodes
		upperBound := geodes + geoR*remaining + remaining*(remaining-1)/2
		if upperBound <= best {
			return
		}

		if time == timeLimit {
			if geodes > best {
				best = geodes
			}
			return
		}

		// Cap resources to max usable
		cappedOre := min(ore, remaining*maxOre)
		cappedClay := min(clay, remaining*maxClay)
		cappedObs := min(obs, remaining*maxObs)

		// State deduplication using packed key
		state := State{time, cappedOre, cappedClay, cappedObs, oreR, clayR, obsR, geoR}
		key := state.key()
		if prev, exists := seen[key]; exists && prev >= geodes {
			return
		}
		seen[key] = geodes

		// Collect resources
		newOre := cappedOre + oreR
		newClay := cappedClay + clayR
		newObs := cappedObs + obsR
		newGeodes := geodes + geoR

		// Try building geode robot (always do if possible - priority)
		if cappedOre >= bp.geoOre && cappedObs >= bp.geoObs {
			dfs(time+1, newOre-bp.geoOre, newClay, newObs-bp.geoObs, newGeodes,
				oreR, clayR, obsR, geoR+1)
			return // If we can build geode robot, always do it
		}

		// Try building obsidian robot
		if cappedOre >= bp.obsOre && cappedClay >= bp.obsClay && obsR < maxObs {
			dfs(time+1, newOre-bp.obsOre, newClay-bp.obsClay, newObs, newGeodes,
				oreR, clayR, obsR+1, geoR)
		}

		// Try building clay robot
		if cappedOre >= bp.clayOre && clayR < maxClay {
			dfs(time+1, newOre-bp.clayOre, newClay, newObs, newGeodes,
				oreR, clayR+1, obsR, geoR)
		}

		// Try building ore robot
		if cappedOre >= bp.oreOre && oreR < maxOre {
			dfs(time+1, newOre-bp.oreOre, newClay, newObs, newGeodes,
				oreR+1, clayR, obsR, geoR)
		}

		// Do nothing (wait)
		dfs(time+1, newOre, newClay, newObs, newGeodes,
			oreR, clayR, obsR, geoR)
	}

	dfs(0, 0, 0, 0, 0, 1, 0, 0, 0)
	return best
}

func part1(blueprints []Blueprint) int {
	total := 0
	for _, bp := range blueprints {
		geodes := maxGeodes(bp, 24)
		total += bp.id * geodes
	}
	return total
}

func part2(blueprints []Blueprint) int {
	result := 1
	count := min(3, len(blueprints))
	for i := 0; i < count; i++ {
		geodes := maxGeodes(blueprints[i], 32)
		result *= geodes
	}
	return result
}

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)
	inputPath := filepath.Join(dir, "..", "input.txt")

	blueprints := parseInput(inputPath)

	fmt.Println("Part 1:", part1(blueprints))
	fmt.Println("Part 2:", part2(blueprints))
}
